src/pudl/transform/eia861.py

```diff
@@ -214,47 +324,495 @@ def balancing_authority(raw_dfs, tfr_dfs):
+
+    Args:
+        raw_dfs (dict): A dictionary of raw EIA 861 dataframes, from the extract step.
+        tfr_dfs (dict): A dictionary of transformed EIA 861 DataFrames, keyed by table
+            name. It will be mutated by this function.
+
+    Returns:
+        dict: A dictionary of transformed EIA 861 dataframes, keyed by table name.
+
+    """
+    return tfr_dfs
+
+
+def demand_response(raw_dfs, tfr_dfs):
+    """
+    Transform the EIA 861 Demand Response table.
+
+    Args:
+        raw_dfs (dict): A dictionary of raw EIA 861 dataframes, from the
+        extract step. tfr_dfs (dict): A dictionary of transformed EIA 861
+        DataFrames, keyed by table name. It will be mutated by this function.
+
+    Returns:
+        dict: A dictionary of transformed EIA 861 dataframes, keyed by table
+            name.
+    """
+    idx_cols = [
+        "utility_id_eia",
+        "state",
+        "report_year",
+        "balancing_authority_code_eia",
+    ]
+
+    customer_classes = [
+        "commercial",
+        "industrial",
+        "other",
+        "residential",
+        "total",
+        "transportation"
+    ]
+
+    ###########################################################################
+    # Tidy Data:
+    ###########################################################################
+    logger.info("Tidying the EIA 861 Demand Response table.")
+    # Clean up values just enough to use primary key columns as a multi-index:
+    logger.debug("Cleaning up EIA861 Demand Response index columns so we can tidy data.")
+    raw_dr = (
+        raw_dfs["demand_response_eia861"].copy()
+        .assign(balancing_authority_code_eia=lambda x: x.balancing_authority_code_eia.fillna("UNK"))
+        .dropna(subset=["utility_id_eia"])
+        .query("utility_id_eia not in (88888, 99999)")
+        .astype({"utility_id_eia": pd.Int64Dtype()})
+        .set_index(idx_cols)
+    )
+    # Split the table into index, data, and "denormalized" columns for processing:
+    # Separate customer classes and reported data into a hierarchical index
+    logger.debug("Stacking EIA861 Demand Response data columns by customer class.")
+    data_cols = _filter_customer_cols(raw_dr, customer_classes)
+    data_cols.columns = (
+        data_cols.columns.str.split("_", n=1, expand=True)
+        .set_names(["customer_class", None])
+    )
+    # Now stack the customer classes into their own categorical column,
+    data_cols = (
+        data_cols.stack(level=0, dropna=False)
+        .reset_index()
+    )
+
+    denorm_cols = _filter_non_customer_cols(
+        raw_dr, customer_classes).reset_index()
+
+    # Merge the index, data, and denormalized columns back together
+    tidy_dr = pd.merge(denorm_cols, data_cols, on=idx_cols)
+
+    # Remove the now redundant "Total" records -- they can be reconstructed
+    # from the other customer classes.
+    tidy_dr = tidy_dr.query("customer_class!='total'")
+    tidy_nrows = len(tidy_dr)
```

Comment:
Did you test whether the total values can be accurately reproduced based on the other reported data? Or find any discrepancies between them that indicated bad data / index values?